(DO NOT MERGE) Tentative `gvfs-2.20.0-rc0`

dscho opened the pull request:
This is a tentative rebase of `gvfs-2.19.1` onto `v2.20.0-rc0.windows.1`.

The majority of the (quite painful) merge conflicts stemmed from the split of `Documentation/config.txt` into `Documentation/config/*.txt`. I am glad that this is over.

Other notable changes:

- The field `commitable` in `struct wt_status` was renamed to `committable` in git.git (6fa9019494f0 (wt-status: rename commitable to committable, 2018-09-05)); I fixed the build by following said rename for that field, but kept the key name in the serialized status as `commitable`, for backwards-compatibility. @jeffhostetler do you agree with this approach?
- The trace2 output of `wt_status_collect()` had merge conflicts because that function was enhanced a little: it now also performs a check whether the current state is committable; I added another trace2 call pair (see ee1aef9d0434b1828cabb3c06e01c90b06c3b9a6, the pre-rebase version is here: 7a6b1ccec1f7dd232d91e2fc16101b2d3e8c83b9). @jeffhostetler do you agree with this change?
- af4ced38b7c2 (commit: add generation to pop_most_recent_commit(), 2018-06-13) had to touch `ref_newer()`, but the post-rebase version (affa5d45adec) seems not to need that, as `ref_newer()` learned to use `is_descendant_of()` in the meantime and no longer uses `pop_most_recent_commit()` (and `ref_newer()` also moved from `remote.c` to `commit-reach.c`). Do you concur with this reasoning, @derrickstolee?
- I see that fbd98c544e4c (fsck: use ERROR_MULTI_PACK_INDEX, 2018-09-24) is still here. @derrickstolee did you mean to send this to the Git mailing list?

A ton of our patches made it upstream (thanks @benpeart, @derrickstolee and @jeffhostetler!!!). To make it easier to review this, I wrote this little shell script (which essentially uses the output of `range-diff` to identify "dropped" commits and then looks for commits with the same oneline that are reachable from the upstream tag):

```sh
git range-diff -s v2.20.0-rc0.windows.1 gvfs-2.19.1 tentative/gvfs-2.20.0-rc0 |
sed -n 's/^ *[0-9]*: *\([0-9a-f]*\)[ -<:]* ----* /\1 /p' |
sed 's/[][()]/\\\\&/g' |
while read hash oneline
do
	ours="$(git show -s --pretty='tformat:%h (%s, %ad)' --date=short $hash)"
	if upstream="$(git show -s --pretty='tformat:%h (%s, %ad)' --date=short v2.20.0-rc0^{/"^$oneline"} 2>/dev/null)"
	then
		echo "* $ours is upstream:<br />$upstream"
	else
		echo "* *NOT* upstream: $ours"
	fi
done
```

The output:

* *NOT* upstream: 7542e0283198 (gvfs: speed up "git checkout -b foo", 2016-09-06)
* 9b8feefac5bc (multi-pack-index: add design document, 2018-07-12) is upstream:<br />ceab693d1f19 (multi-pack-index: add design document, 2018-07-12)
* be46a8a2aa25 (multi-pack-index: add format details, 2018-07-12) is upstream:<br />e0d1bcf82590 (multi-pack-index: add format details, 2018-07-12)
* fa73d892e385 (multi-pack-index: add builtin, 2018-07-12) is upstream:<br />6a257f03ba9b (multi-pack-index: add builtin, 2018-07-12)
* c235d587c811 (multi-pack-index: add 'write' verb, 2018-07-12) is upstream:<br />a3407730261b (multi-pack-index: add 'write' verb, 2018-07-12)
* 0143fbfceed4 (midx: write header information to lockfile, 2018-07-12) is upstream:<br />fc59e7484461 (midx: write header information to lockfile, 2018-07-12)
* dbb9756fbc91 (multi-pack-index: load into memory, 2018-07-12) is upstream:<br />4d80560c5461 (multi-pack-index: load into memory, 2018-07-12)
* 602f50fc8dd0 (t5319: expand test data, 2018-07-12) is upstream:<br />2c3813354b2c (t5319: expand test data, 2018-07-12)
* 2e586367bd13 (packfile: generalize pack directory list, 2018-07-12) is upstream:<br />9208e318f548 (packfile: generalize pack directory list, 2018-07-12)
* a8b537185775 (multi-pack-index: read packfile list, 2018-07-12) is upstream:<br />396f257018a6 (multi-pack-index: read packfile list, 2018-07-12)
* 1a05267c0e7d (multi-pack-index: write pack names in chunk, 2018-07-12) is upstream:<br />32f3c541e3c8 (multi-pack-index: write pack names in chunk, 2018-07-12)
* 72f7c3015c15 (midx: read pack names into array, 2018-07-12) is upstream:<br />3227565cfdea (midx: read pack names into array, 2018-07-12)
* dd95da86e0fe (midx: sort and deduplicate objects from packfiles, 2018-07-12) is upstream:<br />fe1ed56f5e48 (midx: sort and deduplicate objects from packfiles, 2018-07-12)
* afddbab5bae9 (midx: write object ids in a chunk, 2018-07-12) is upstream:<br />0d5b3a5ef723 (midx: write object ids in a chunk, 2018-07-12)
* 6433f018fa2e (midx: write object id fanout chunk, 2018-07-12) is upstream:<br />d7cacf29ccfc (midx: write object id fanout chunk, 2018-07-12)
* 1ad3fbbe88b8 (midx: write object offsets, 2018-07-12) is upstream:<br />662148c43566 (midx: write object offsets, 2018-07-12)
* e065fe415108 (config: create core.multiPackIndex setting, 2018-07-12) is upstream:<br />c4d25228ebb2 (config: create core.multiPackIndex setting, 2018-07-12)
* f581c575916e (midx: read objects from multi-pack-index, 2018-07-12) is upstream:<br />3715a6335c37 (midx: read objects from multi-pack-index, 2018-07-12)
* 0bc9c2e1fa90 (midx: use midx in abbreviation calculations, 2018-07-12) is upstream:<br />8aac67a17406 (midx: use midx in abbreviation calculations, 2018-07-12)
* d6b2eb3a8735 (midx: use existing midx when writing new one, 2018-07-12) is upstream:<br />a40498a12654 (midx: use existing midx when writing new one, 2018-07-12)
* 8b859754fde6 (midx: use midx in approximate_object_count, 2018-07-12) is upstream:<br />b8990fbfedf7 (midx: use midx in approximate_object_count, 2018-07-12)
* 3d6df6b68377 (midx: prevent duplicate packfile loads, 2018-07-12) is upstream:<br />f3a002bd8479 (midx: prevent duplicate packfile loads, 2018-07-12)
* bffccde8eb54 (packfile: skip loading index if in multi-pack-index, 2018-07-12) is upstream:<br />17c35c89698c (packfile: skip loading index if in multi-pack-index, 2018-07-12)
* 8d1776d1b2ac (midx: clear midx on repack, 2018-07-12) is upstream:<br />525e18c04bb3 (midx: clear midx on repack, 2018-07-12)
* 752b7d40bdeb (multi-pack-index: provide more helpful usage info, 2018-08-20) is upstream:<br />6d68e6a46174 (multi-pack-index: provide more helpful usage info, 2018-08-20)
* 6cc4f2ef400e (multi-pack-index: store local property, 2018-08-20) is upstream:<br />2cf489a3bf75 (multi-pack-index: store local property, 2018-08-20)
* 5feb5457d98c (midx: mark bad packed objects, 2018-08-20) is upstream:<br />c39b02ae0ae9 (midx: mark bad packed objects, 2018-08-20)
* 4adca2655deb (midx: stop reporting garbage, 2018-08-20) is upstream:<br />fe86c3beb589 (midx: stop reporting garbage, 2018-08-20)
* 9145f1b05c36 (midx: fix bug that skips midx with alternates, 2018-08-20) is upstream:<br />29e2016b8f95 (midx: fix bug that skips midx with alternates, 2018-08-20)
* ff1aba769a3a (packfile: add all_packs list, 2018-08-20) is upstream:<br />0bff5269d3ed (packfile: add all_packs list, 2018-08-20)
* c15c0d5c75fa (treewide: use get_all_packs, 2018-08-20) is upstream:<br />454ea2e4d703 (treewide: use get_all_packs, 2018-08-20)
* 3c34a969ed81 (midx: test a few commands that use get_all_packs, 2018-08-20) is upstream:<br />e9ab2ed7de33 (midx: test a few commands that use get_all_packs, 2018-08-20)
* acf68eb7defe (pack-objects: consider packs in multi-pack-index, 2018-08-20) is upstream:<br />6a22d521260f (pack-objects: consider packs in multi-pack-index, 2018-08-20)
* 1c47a55ae960 (multi-pack-index: add 'verify' verb, 2018-09-13) is upstream:<br />56ee7ff15655 (multi-pack-index: add 'verify' verb, 2018-09-13)
* ca8f1b15b45c (multi-pack-index: verify bad header, 2018-09-13) is upstream:<br />53ad0407444a (multi-pack-index: verify bad header, 2018-09-13)
* b0e147a980f2 (multi-pack-index: verify corrupt chunk lookup table, 2018-09-13) is upstream:<br />d3f8e211700c (multi-pack-index: verify corrupt chunk lookup table, 2018-09-13)
* ac57e43cbcce (multi-pack-index: verify packname order, 2018-09-13) is upstream:<br />8e72a3c321eb (multi-pack-index: verify packname order, 2018-09-13)
* 52cbaf81b61e (multi-pack-index: verify missing pack, 2018-09-13) is upstream:<br />d4bf1d88b902 (multi-pack-index: verify missing pack, 2018-09-13)
* 43ca741db549 (multi-pack-index: verify oid fanout order, 2018-09-13) is upstream:<br />2f23d3f3f92d (multi-pack-index: verify oid fanout order, 2018-09-13)
* a50401646759 (multi-pack-index: verify oid lookup order, 2018-09-13) is upstream:<br />55c5648d804c (multi-pack-index: verify oid lookup order, 2018-09-13)
* b99eab2cb6fb (multi-pack-index: fix 32-bit vs 64-bit size check, 2018-09-13) is upstream:<br />d8ac9ee109ad (multi-pack-index: fix 32-bit vs 64-bit size check, 2018-09-13)
* c7dec4cf9a19 (multi-pack-index: verify object offsets, 2018-09-13) is upstream:<br />cc6af73c029d (multi-pack-index: verify object offsets, 2018-09-13)
* 505e8cdc3304 (multi-pack-index: report progress during 'verify', 2018-09-13) is upstream:<br />144d70333e82 (multi-pack-index: report progress during 'verify', 2018-09-13)
* d3503948a6b9 (fsck: verify multi-pack-index, 2018-09-13) is upstream:<br />66ec0390e754 (fsck: verify multi-pack-index, 2018-09-13)
* 896bea5c434b (midx: fix broken free() in close_midx(), 2018-10-08) is upstream:<br />0ce4ff942125 (midx: fix broken free() in close_midx(), 2018-10-08)
* 6f2c87a1dfd8 (midx: close multi-pack-index on repack, 2018-10-08) is upstream:<br />1dcd9f2043a3 (midx: close multi-pack-index on repack, 2018-10-12)
* 9d0f2de5d921 (multi-pack-index: define GIT_TEST_MULTI_PACK_INDEX, 2018-08-29) is upstream:<br />0465a5050602 (multi-pack-index: define GIT_TEST_MULTI_PACK_INDEX, 2018-10-12)
* 397b80fde6e7 (trace.h: support nested performance tracing, 2018-08-18) is upstream:<br />c46c406ae1ee (trace.h: support nested performance tracing, 2018-08-18)
* 1d9ed82b38ef (unpack-trees: add performance tracing, 2018-08-18) is upstream:<br />0d1ed5963da5 (unpack-trees: add performance tracing, 2018-08-18)
* 73bcf0f4f2fb (unpack-trees: optimize walking same trees with cache-tree, 2018-08-18) is upstream:<br />b4da37380b77 (unpack-trees: optimize walking same trees with cache-tree, 2018-08-18)
* 18bc2882e09e (unpack-trees: reduce malloc in cache-tree walk, 2018-08-18) is upstream:<br />f1e11c651084 (unpack-trees: reduce malloc in cache-tree walk, 2018-08-18)
* e253d3f1ea54 (unpack-trees: reuse (still valid) cache-tree from src_index, 2018-08-18) is upstream:<br />836ef2b69f3a (unpack-trees: reuse (still valid) cache-tree from src_index, 2018-08-18)
* 07fc8cf01f43 (unpack-trees: add missing cache invalidation, 2018-08-18) is upstream:<br />5697ca9aa562 (unpack-trees: add missing cache invalidation, 2018-08-18)
* 0637198f04f0 (cache-tree: verify valid cache-tree in the test suite, 2018-08-18) is upstream:<br />4592e6080ff0 (cache-tree: verify valid cache-tree in the test suite, 2018-08-18)
* ba708462ab5f (Document update for nd/unpack-trees-with-cache-tree, 2018-08-25) is upstream:<br />5f4436a7211d (Document update for nd/unpack-trees-with-cache-tree, 2018-08-25)
* 4737e38ee602 (t0051: test GIT_TRACE to a windows named pipe, 2018-09-07) is upstream:<br />06ba9d03e34f (t0051: test GIT_TRACE to a windows named pipe, 2018-09-11)
* 483c8d54e0e3 (mingw: fix mingw_open_append to work with named pipes, 2018-09-06) is upstream:<br />eeaf7ddac759 (mingw: fix mingw_open_append to work with named pipes, 2018-09-11)
* *NOT* upstream: 9d4b3939630e (Rename one of the t5319 tests, 2018-10-25)

The first of the two *NOT* upstream commits is of course a false positive: fa655d8411cc (checkout: optimize "git checkout -b <new_branch>", 2018-08-16) *is* upstream. @benpeart do we have to do anything to put that back under the `core.gvfs` flag? If I read this correctly, then the speed-up is hidden behind a config setting that GVFS does *not* set, right?

The second *NOT* upstream commit is obsolete, as Git for Windows renamed t5319 to t5321 (not t5320).

derrickstolee commented:
@dscho Sorry about the `ref_newer()` change. Yes, we should take what upstream has. The peformance is the same, but it came along with a bigger refactoring in the upstream version. We should just drop the commits that were changing it here.

derrickstolee commented on Documentation/git-apply.txt:
```diff
@@ -66,7 +66,7 @@ OPTIONS
 	applicable to what the current index file records.  If
 	the file to be patched in the working tree is not
-	up to date, it is flagged as an error.  This flag also
+	up-to-date, it is flagged as an error.  This flag also
```
Are we ready to drop this "up-to-date" change? It had an interaction with VS. @dscho @jrbriggs ?

mjcheetham replied:
We updated both VS 15.0 and 15.7+ with the fix to accept both "Already up-to-date" or "Already up to date" from the output of `git-merge` and `git-pull`.

dscho replied:
I opened a PR to drop this Revert: https://github.com/Microsoft/git/pull/72

derrickstolee commented:
Build is failing on `t7524-serialized-status.sh` for Linux and Mac, but not Windows. @jamill can you take a look?

dscho commented:
@kewillford could you have a look why https://gvfs.visualstudio.com/ci/_build/results?buildId=4066&view=logs *passes* in the Windows phase, despite there clearly being a problem in `t0000-basic.sh`?

dscho commented:
> could you have a look why https://gvfs.visualstudio.com/ci/_build/results?buildId=4066&view=logs _passes_ in the Windows phase

Actually, I think I got this. I simply added

```powershell
if (!$?) {
    exit(1)
}
```

after the call

```powershell
C:\git-sdk-64-ci\git-cmd.exe --command=usr\bin\sh.exe -lc @'
set -e
make -j10 DEVELOPER=1

make GIT_TEST_OPTS=--quiet -j10 -C t ||
make GIT_TEST_OPTS=\"-i -v -x\" -k -C t failed 
'@
```

dscho commented:
@derrickstolee I already had a look, and added 6657af1814. Hopefully this fixes it the correct way. @jeffhostetler could you review that fixup! commit and tell me whether it is correct?

dscho commented:
> @kewillford could you have a look why https://gvfs.visualstudio.com/ci/_build/results?buildId=4066&view=logs _passes_

@kewillford I also updated the build definition to use the Ubuntu1604 pool (Hosted Linux Preview is deprecated and will go away soon, and the only additional change that is required is to use `sudo` with `apt-get`, as HLP was container-based and Ubuntu1604 is VM-based).

dscho commented:
> Hosted Linux Preview is deprecated

@kewillford see https://aka.ms/linux-preview-deprecation for details; you may want to inspect your definitions before Dec 1st. The easiest way, I've been told, is to go to the admin page and look for the "Hosted Linux Preview" pool, then click on all of the agents in that pool to see which builds they ran lately.

dscho commented:
> I also updated the build definition to use the Ubuntu1604 pool

@kewillford I also updated the Windows phase to run on the "Hosted" pool instead of "Hosted VS2017". In my hands, the former performs some 20% faster than the latter.

jeffhostetler commented on wt-status-deserialize.c:
```diff
@@ -812,8 +812,11 @@ int wt_status_deserialize(const struct wt_status *cmd_s,
 
-	if (result == DESERIALIZE_OK)
+	if (result == DESERIALIZE_OK) {
+		wt_status_get_state(&des_s.state, des_s.branch &&
```
This looks fine.

jeffhostetler commented:
@dscho WRT "commitable" vs "committable":  either way is fine. new deserialize clients will reject
status-cache files from old serialize clients, and fall back and run a real status.  Since status-cache
files are very short-lived (they are invalidated any time anything changes in the worktree), so it
really won't matter.

I'd say change it to "committable" to be consistent with the upstream change.

jeffhostetler commented:
@dscho  WRT the trace2 in wt_status_collect().  that looks fine. thanks.

kewillford commented:
@dscho changing the pool to use Ubuntu failed.  https://gvfs.visualstudio.com/ci/_build/results?buildId=4084&_a=summary&view=logs 
```
2018-11-21T16:41:37.4132217Z Reading package lists...
2018-11-21T16:41:37.4406105Z W: chmod 0700 of directory /var/lib/apt/lists/partial failed - SetupAPTPartialDirectory (1: Operation not permitted)
2018-11-21T16:41:37.4407293Z E: Could not open lock file /var/lib/apt/lists/lock - open (13: Permission denied)
2018-11-21T16:41:37.4407676Z E: Unable to lock directory /var/lib/apt/lists/
2018-11-21T16:41:37.4408580Z W: Problem unlinking the file /var/cache/apt/pkgcache.bin - RemoveCaches (13: Permission denied)
2018-11-21T16:41:37.4438740Z W: Problem unlinking the file /var/cache/apt/srcpkgcache.bin - RemoveCaches (13: Permission denied)
```

dscho commented:
On Wed, 21 Nov 2018, Kevin Willford wrote:

> @dscho changing the pool to use Ubuntu failed.  https://gvfs.visualstudio.com/ci/_build/results?buildId=4084&_a=summary&view=logs 
> ```
> 2018-11-21T16:41:37.4132217Z Reading package lists...
> 2018-11-21T16:41:37.4406105Z W: chmod 0700 of directory /var/lib/apt/lists/partial failed - SetupAPTPartialDirectory (1: Operation not permitted)
> 2018-11-21T16:41:37.4407293Z E: Could not open lock file /var/lib/apt/lists/lock - open (13: Permission denied)
> 2018-11-21T16:41:37.4407676Z E: Unable to lock directory /var/lib/apt/lists/
> 2018-11-21T16:41:37.4408580Z W: Problem unlinking the file /var/cache/apt/pkgcache.bin - RemoveCaches (13: Permission denied)
> 2018-11-21T16:41:37.4438740Z W: Problem unlinking the file /var/cache/apt/srcpkgcache.bin - RemoveCaches (13: Permission denied)
> ```

I see you fixed it already by prefixing the `apt-get` calls with `sudo`.  That is a documented thing: Hosted Linux Preview was running inside Containers (where `sudo` is not necessary), but Ubuntu 16.04 runs on real VMs.
